proposals/79/flink_integration.md

```diff
@@ -0,0 +1,144 @@
+
+There is one caveat, inside `execute()` method call, there is a
+call: `final StreamGraph streamGraph = getStreamGraph();`, which is calling `getStreamGraph(true)` internally. 
+That method with parameter `getStreamGraph(boolean clearTransformations)`:
+
+```java
+@Public
+public class StreamExecutionEnvironment {
+    /*
+     * ......
+     * .....
+     */
+    /**
+     * Getter of the {@link StreamGraph} of the streaming job with the option to clear previously
+     * registered {@link Transformation transformations}. Clearing the transformations allows, for
+     * example, to not re-execute the same operations when calling {@link #execute()} multiple
+     * times.
+     *
+     * @param clearTransformations Whether or not to clear previously registered transformations
+     * @return The stream graph representing the transformations
+     */
+    @Internal
+    public StreamGraph getStreamGraph(boolean clearTransformations) {
+        final StreamGraph streamGraph = getStreamGraphGenerator(transformations).generate();
+        if (clearTransformations) {
+            transformations.clear();
+        }
+        return streamGraph;
+    }
+}
+```
+clears out the `List<Transformation<?>> transformations = new ArrayList<>()`.<br/>
+As this call is happening before `JobListener.onJobSubmitted()` method call, we can't get pipeline/streamGraph out of Flink.<br/>
+`StreamGraph` is constructed based on this list of `transformation`s.<br/>
+
+To address this issue we are now tweaking temporarily with reflection the field `transformations` inside `StreamExecutionEnvironment` class:
+```java
+public class OpenLineageFlinkJobListener implements JobListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(OpenLineageFlinkJobListener.class);
+    private final StreamExecutionEnvironment executionEnvironment;
+
+    static class UnclearableList<T> extends ArrayList<T> {
+        public UnclearableList(Collection<T> collection) {super(collection);}
+        @Override
+        public void clear() {LOGGER.info("Ignore clearing the transformations ArrayList");}
+    }
+
+    OpenLineageFlinkJobListener(StreamExecutionEnvironment executionEnvironment) {
+        this.executionEnvironment = executionEnvironment;
+        try {
+            Field transformations = FieldUtils.getField(StreamExecutionEnvironment.class, "transformations", true);
+            ArrayList<Transformation<?>> previousTransformationList = (ArrayList<Transformation<?>>) FieldUtils.readField(transformations, executionEnvironment, true);
+            List<Transformation<?>> transformationList = new UnclearableList<>(previousTransformationList);
+            FieldUtils.writeField(transformations, executionEnvironment, transformationList, true);
+        } catch (IllegalAccessException e) {
+            LOGGER.error("Failed to rewrite transformations");
+        }
+    }
```

Comment:
Yuck 🤢

(I know you feel the same 😜)

Can you not call the version with the boolean set to `false`?

Reply:
It was introduced here: https://issues.apache.org/jira/browse/FLINK-15093
Reason is explained in the Flink codebase:
```
/**
     * Getter of the {@link StreamGraph} of the streaming job with the option to clear previously
     * registered {@link Transformation transformations}. Clearing the transformations allows, for
     * example, to not re-execute the same operations when calling {@link #execute()} multiple
     * times.
     *
     * @param clearTransformations Whether or not to clear previously registered transformations
     * @return The stream graph representing the transformations
     */
    @Internal
    public StreamGraph getStreamGraph(boolean clearTransformations) {
        final StreamGraph streamGraph = getStreamGraphGenerator(transformations).generate();
        if (clearTransformations) {
            transformations.clear();
        }
        return streamGraph;
    }
```
https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L2134

we can't call the version of this method with `false` as it is hardcoded in `execute()` method of `StreamExecutionEnvironment`, which is the entry-point for Flink applications (loads `transformation`s and generates `StreamGraph`)